turbopack/crates/turbo-tasks-backend/src/utils/dash_map_multi.rs

```diff
@@ -0,0 +1,178 @@
+        }
+    }
+}
+
+impl<'a, K: Eq + Hash, V, S: BuildHasher> Deref for RefMut<'a, K, V, S> {
+    type Target = V;
+
+    fn deref(&self) -> &V {
+        self.value()
+    }
+}
+
+impl<'a, K: Eq + Hash, V, S: BuildHasher> DerefMut for RefMut<'a, K, V, S> {
+    fn deref_mut(&mut self) -> &mut V {
+        self.value_mut()
+    }
+}
+
+impl<'a, K, V, S> From<dashmap::mapref::one::RefMut<'a, K, V, S>> for RefMut<'a, K, V, S>
+where
+    K: Hash + Eq,
+    S: BuildHasher,
+{
+    fn from(r: dashmap::mapref::one::RefMut<'a, K, V, S>) -> Self {
+        RefMut::Base(r)
+    }
+}
+
+pub fn get_multiple_mut<K, V, S>(
+    map: &DashMap<K, V, S>,
+    key1: K,
+    key2: K,
+    insert_with: impl Fn() -> V,
+) -> (RefMut<'_, K, V, S>, RefMut<'_, K, V, S>)
+where
+    K: Hash + Eq + Clone,
+    S: BuildHasher + Clone,
+{
+    let s1 = map.determine_map(&key1);
+    let s2 = map.determine_map(&key2);
+    let shards = map.shards();
+    if s1 == s2 {
+        let mut guard = shards[s1].write();
+        let e1 = guard
+            .raw_entry_mut()
+            .from_key(&key1)
+            .or_insert_with(|| (key1.clone(), SharedValue::new(insert_with())));
+        let mut key1_ptr = e1.0 as *const K;
+        let mut value1_ptr = e1.1.get_mut() as *mut V;
+        let key2_ptr;
+        let value2_ptr;
+        match guard.raw_entry_mut().from_key(&key2) {
+            hash_map::RawEntryMut::Occupied(e) => {
+                let e2 = e.into_key_value();
+                key2_ptr = e2.0 as *const K;
+                value2_ptr = e2.1.get_mut() as *mut V;
+            }
+            hash_map::RawEntryMut::Vacant(e) => {
+                let e2 = e.insert(key2.clone(), SharedValue::new(insert_with()));
+                key2_ptr = e2.0 as *const K;
+                value2_ptr = e2.1.get_mut() as *mut V;
+                // inserting a new entry might invalidate the pointers of the first entry
+                let e1 = guard.get_key_value_mut(&key1).unwrap();
+                key1_ptr = e1.0 as *const K;
+                value1_ptr = e1.1.get_mut() as *mut V;
+            }
+        }
+        let guard = Arc::new(guard);
+        (
+            RefMut::Shared {
+                guard: guard.clone(),
+                key: key1_ptr,
+                value: value1_ptr,
+            },
+            RefMut::Shared {
+                guard,
+                key: key2_ptr,
+                value: value2_ptr,
+            },
+        )
+    } else {
+        let (mut guard1, mut guard2) = loop {
+            {
+                let g1 = shards[s1].write();
+                if let Some(g2) = shards[s2].try_write() {
+                    break (g1, g2);
+                }
+            }
+            {
+                let g2 = shards[s2].write();
+                if let Some(g1) = shards[s1].try_write() {
+                    break (g1, g2);
+                }
+            }
+        };
```

Comment:
I'm pretty sure you can avoid the loop if you just ensure that you access shards in a deterministic order to avoid lock inversion. That can be as simple as accessing the shard with the smaller index first.

Reply:
I think you are right, but would this be better?

With the ordered approach it could hold the smaller lock while waiting for the bigger lock. So some time unnecessary holding a lock.

Example

A wants 3 and 4, B wants 2 and 3 and C wants 1 and 2.

They are executed in order A, B, C all needing to operate for 1ms.

Ordered:
0ms:
A locks 3 and 4 and starts doing something.
B locks 2 and waits for 3 as it's already locked.
C locks 1 and waits for 2 as it's already locked.
1ms:
A unlocks 3 and 4.
B gets 3 and starts doing something.
2ms:
B unlocks 2 and 3.
C gets 2 and starts doing something.
3ms:
C unlocks 1 and 2.
Finished after 3ms.

Current:
0ms:
A locks 3 and 4 and starts doing something.
B locks 2, doesn't get 3, unlocks 2 and waits for 3.
C locks 1 and 2 and starts doing something.
1ms:
A unlocks 3 and 4.
C unlocks 1 and 2.
B gets 3, locks 2 and starts doing something.
2ms:
B unlocks 2 and 3.
FInished after 2ms.

Reply:
You're right. It's really hard to say what's better.

I think the core problem with a deterministic ordering based on index order is that the earlier shard rwlocks will experience higher lock contention than the later rwlocks.

Hopefully this shouldn't really matter though, as the `DashMap` should have a sufficient number of shards. If it doesn't, we should just increase the number of shards.